### ISSUE TITLE
[MASTER FEATURE] New default loaders for AMP

### ISSUE THREAD

**Issue by nainar** [ampproject/amphtml#20237]
New loaders that indicate an image or embedded feature is loading will be utilized in AMP and are in the process of being built. A current 3-dot loader exists. This issue also encompasses the research  necessary to determine whether redesigns will be received favorably by comparison.

cc @aghassemi @andrewwatterson

**Comment by nainar** [comment]
Context here: #3830

**Comment by nainar** [comment]
## Loaders design principles

- Based on User Research Studies, we noticed that staged loaders (loaders that can be broken up into discrete stages) performed the best on the metrics of user happiness and their perception of how long the loader takes.
- Based on this insight we designed stage loaders for All components that make a server request to initialize.  
- The loaders are split into 3 stages: nothing, the context loads (Ad, Facebook logo, etc) and lastly the spinner spins round the central badge. 

## Loaders proposal

### Generic loader
**Small**
![default-small-0](https://user-images.githubusercontent.com/7919520/54077290-320a9900-426b-11e9-8726-5a246e2238b4.gif)
Without image placeholder
![default-small-1](https://user-images.githubusercontent.com/7919520/54077289-320a9900-426b-11e9-9c0c-c85f79503726.gif)
With image placeholder

**Default size**
![default-default-0](https://user-images.githubusercontent.com/7919520/54077292-320a9900-426b-11e9-8d52-afbe475a4db9.gif)
Without image placeholder
![default-default-1](https://user-images.githubusercontent.com/7919520/54077291-320a9900-426b-11e9-821e-cd81aef647d6.gif)
With image placeholder

### Content specific loader
#### Video loader
![video-default-0](https://user-images.githubusercontent.com/7919520/54077280-30d96c00-426b-11e9-9117-e9fef2c049d2.gif)
Without image placeholder
![video-default-1](https://user-images.githubusercontent.com/7919520/54077279-30d96c00-426b-11e9-9858-43c96e6db041.gif)
With image placeholder

#### Ads loader
![ad-default-0](https://user-images.githubusercontent.com/7919520/54077295-32a32f80-426b-11e9-9f5c-a169732c3fa0.gif)
Without image placeholder
![ad-default-1](https://user-images.githubusercontent.com/7919520/54077293-320a9900-426b-11e9-9dfa-dd5300891f65.gif)
With image placeholder

#### Twitter loader
![twitter-default-0](https://user-images.githubusercontent.com/7919520/54077282-30d96c00-426b-11e9-8e9f-0522b1a62328.gif)
Without image placeholder
![twitter-default-1](https://user-images.githubusercontent.com/7919520/54077281-30d96c00-426b-11e9-972a-b8b7d8c9126f.gif)
With image placeholder

#### Pinterest loader
![pinterest-default-0](https://user-images.githubusercontent.com/7919520/54077284-31720280-426b-11e9-9d8d-4d1d6527e7e3.gif)
Without image placeholder
![pinterest-default-1](https://user-images.githubusercontent.com/7919520/54077283-31720280-426b-11e9-89b2-5d1c20a209c6.gif)
With image placeholder

#### Instagram loader
![instagram-large-0](https://user-images.githubusercontent.com/7919520/54077286-31720280-426b-11e9-92ad-665e9e903a6e.gif)
Without image placeholder
![instagram-large-1](https://user-images.githubusercontent.com/7919520/54077285-31720280-426b-11e9-8542-43f15603c3af.gif)
With image placeholder

#### Facebook loader
![facebook-default-0](https://user-images.githubusercontent.com/7919520/54077288-31720280-426b-11e9-858b-33dc45fec353.gif)
Without image placeholder
![facebook-default-1](https://user-images.githubusercontent.com/7919520/54077287-31720280-426b-11e9-9bab-1c2cd9193151.gif)
With image placeholder

cc @andrewwatterson @aghassemi 

Paging @cramforce @rudygalfi @cpapazian @paularmstrong

**Comment by cramforce** [comment]
I ❤️ all of these! Can't wait to get rid of the three dots.

**Comment by yazdotdev** [comment]
Yes @cramforce !  Goodbye three dots

**Comment by yazdotdev** [comment]
@nainar  What would be the recommended placeholder image format (as opposed to fallback)?
WebP does not seem to have progressive loading.  But MozJPEG does.  Which should take priority?

**Comment by yazdotdev** [comment]
@nainar would it be possible to have a custom icon for the loaders?

**Comment by nainar** [comment]
Hi @yazdotdev that's a great idea! To clarify that I understand you correctly, you want to be able to override the default loader with one of your own so that for e.g. an entire site uses the same loader?

**Comment by yazdotdev** [comment]
Hi @nainar thanks for replying. That's pretty much it. 
*If you could :*
- have your own loader with your own logo
- use it as standard on your site (e.g. even if you are not embedded some third party site's content)
- if some other site would like to use your site's content (e.g. embed your content)
    - you could provide the amp code for easy integration
    - it does not even have to be all that different from your standard code (except for some attribution links etc that could just be turned on)  
- please *do not hesitate* to ask me to clarify anything further

**Comment by choumx** [comment]
I'm curious about the when the "image placeholder" variant will be shown, since AMP placeholders are not necessarily images.

**Comment by andrewwatterson** [comment]
Hey @choumx - the plan is not to show loaders over placeholders that are anything other than just an image.

Our observation is that placeholders that are just images tend to be previews of the content (video poster frame, Instagram source image, et cetera), whereas placeholders that have more content than that could be trying to implement either a custom loader or some other experience we can't necessarily assume should have the stock loader over it.

See attached for the full spec that includes a note on that.

![amp-loaders](https://user-images.githubusercontent.com/1370189/55413194-b45e4400-551d-11e9-918c-6d8dabe13f2f.png)
.

**Comment by andrewwatterson** [comment]
Note the above spec shows a small ads loader that we've since decided against implementing. Even 320x50 ad units will get the full-sized ads loader. Ads with an image placeholder will get no loader (the current behavior).

**Comment by jridgewell** [comment]
Do we need to do anything to get this to work with blurry image placeholders, eg https://output.jsbin.com/lokowoj/3/quiet (turn on network throttling to see full effect)?

**Comment by aghassemi** [comment]
@jridgewell yes but it is part of my test plan. no loader will show up with blurry image placeholders.

**Comment by nyedidikeke** [comment]
> Hey @choumx - the plan is not to show loaders over placeholders that are anything other than just an image.
> 
> Our observation is that placeholders that are just images tend to be previews of the content (video poster frame, Instagram source image, et cetera), whereas placeholders that have more content than that could be trying to implement either a custom loader or some other experience we can't necessarily assume should have the stock loader over it.
> 
> See attached for the full spec that includes a note on that.
> 
> ![amp-loaders](https://user-images.githubusercontent.com/1370189/55413194-b45e4400-551d-11e9-918c-6d8dabe13f2f.png)
> .

Kindly share URL to the full spec page.

**Comment by nainar** [comment]
Closing as this has launched. Please fill issues regarding this feature as a new issue.
